src/category_theory/bicategory/coherence.lean

```diff
@@ -0,0 +1,291 @@
+  case whisker_right : _ _ _ _ _ _ _ ih { funext, apply congr_arg2 _ rfl (congr ih rfl) },
+  all_goals { funext, refl }
+end
+
+/--
+Auxiliary definition for `normalize`. Given a 2-morphism between `f` and `g` in the free
+bicategory, we have a natural transformation between `normalize_hom f` and `normalize_hom g`
+that are viewed as functors between discrete categories.
+-/
+@[simps]
+def normalize_map_aux {a b c : B} {f g : hom b c} (η : hom₂ f g) :
+  (discrete.functor (normalize_hom f) : _ ⥤ discrete (path.{v+1} a c)) ⟶
+    discrete.functor (normalize_hom g) :=
+discrete.nat_trans (λ p, eq_to_hom (normalize_hom_congr η p))
+
+/--
+The normalization of the composition of `p : path a b` and `f : hom b c` as a functor.
+-/
+def normalize (a b c : B) : hom b c ⥤ discrete (path.{v+1} a b) ⥤ discrete (path.{v+1} a c) :=
+{ obj := λ f, discrete.functor (normalize_hom f),
+  map := λ f g, quot.lift normalize_map_aux (by tidy) }
+
+/--
+A variant of the normalization functor where we consider the result as a 1-morphism in the free
+bicategory rather than a path.
+-/
+def normalize' (a b c : B) : hom b c ⥤ discrete (path.{v+1} a b) ⥤ hom a c :=
+normalize _ _ _ ⋙ (whiskering_right _ _ _).obj (inclusion_path _ _)
+
+variables (B)
+
+/-- The normalization pseudofunctor for the free bicategory on a quiver `B`. -/
+def full_normalize : pseudofunctor (free_bicategory B) (locally_discrete (paths B)) :=
+{ obj := id,
+  map := λ a b f, ((normalize _ _ _).obj f).obj nil,
+  map₂ := λ a b f g η, ((normalize _ _ _).map η).app nil,
+  map_id := λ a, iso.refl (𝟙 a),
+  map_comp := λ a b c f g, eq_to_iso
+  begin
+    induction g generalizing a,
+    case id { refl },
+    case of { refl },
+    case comp : _ _ _ g _ ihf ihg { erw [ihg _ (f.comp g), ihf _ f, ihg _ g, assoc] }
+  end }
+
+variables {B}
+
+/--
+Given a 1-morphism `f : hom b c` in the free bicategory and a path `p : path a b`, taking the
+composition of `p` and `f` in the free bicategory is functorial in both `f` and `p`.
+-/
+def whisker_path (a b c : B) : hom b c ⥤ discrete (path.{v+1} a b) ⥤ hom a c :=
+{ obj := λ f, discrete.functor (λ p, (preinclusion _).map p ≫ f),
+  map := λ f g η, discrete.nat_trans (λ p, (preinclusion _).map p ◁ η) }
+
+lemma whisker_path_obj_map
+  (a : B) {b c : B} (f : hom b c) {p p' : discrete (path.{v+1} a b)} (η : p ⟶ p') :
+  ((whisker_path _ _ _).obj f).map η = (inclusion_path _ _).map η ▷ f :=
+by tidy
+
+/--
+Auxiliary definition for `normalize_iso`. Here we construct the isomorphism between
+`inclusion_path_aux p ≫ f` and `inclusion_path_aux (normalize_hom f p)`.
+-/
+@[simp]
+def normalize_iso_app {a : B} : Π {b c : B} (f : hom b c) (p : path a b),
+  ((whisker_path a b c).obj f).obj p ≅ ((normalize' a b c).obj f).obj p
+| _ _ (hom.of f) p := iso.refl _
+| _ _ (hom.id a) p := ρ_ ((preinclusion _).map p)
+| _ _ (hom.comp f g) p :=
+    (α_ _ _ _).symm ≪≫ whisker_right_iso (normalize_iso_app f p) g ≪≫
+      normalize_iso_app g (((normalize _ _ _).obj f).obj p)
```

Comment:
Seems the difficulty lies here; this is basically p(fg) -> (pf)g -> (norm(pf))g -> norm(norm(pf)g) = norm(p(fg)). Lean doesn't know that pf is simpler than p(fg), but once we separate p from fg, then Lean knows the f in pf and the g in norm(pf)g are both simpler than the fg in p(fg).